Add a Docker Image for testing and doc building in an isolated environment

PGijsbers opened the pull request:
Due to my recent difficulties building the docs from my Windows machine (still not sure why it broke), I decided to make a docker image that can be used to:
 - run the latest openml-python development version. Note: 'latest' at the time of building, this PR does not yet include the automation to rebuild and push the docker image on each push to `develop`.
 - run unit tests on your local repository (mounted), or any online branch (on this repository or a fork)
 - build documentation on your local repository (mounted),  or any online branch (on this repository or a fork)
 
Please refer to the included documentation for more information.
I think it'll be useful for anyone who wishes to execute these tasks in an isolated environment, though the primary audience is probably Windows users. 

Looking for feedback, but I will still add:
 - [x] A link to the readme from the documentation.
 - [x] A note in the progress log.

mfeurer commented on docker/readme.md:
```diff
@@ -0,0 +1,78 @@
+This section contains some notes about the structure of the image, intended for those who want to work on it.
+
+### Added Directories
+There is a `/omlp` directory which by default contains the openml-python repository in the state with which the image was built.
```
Should these be bullet points?

PGijsbers replied:
That would be an improvement :)

mfeurer commented on docker/startup.sh:
```diff
@@ -0,0 +1,75 @@
+if ! [ -d "/code" ] && [ -z "$2" ]; then
+  echo "To perform $1 a code repository must be mounted to '/code' or a branch must be specified." >> /dev/stderr
+  exit 1
+fi
```
why not elif?

PGijsbers replied:
It doesn't really matter as `exit` is called if the body of any if-statement is entered.
I find this a bit more clear because you can reason about each statement individually without taking into account previous statements (it also indicates the order does not matter).

mfeurer replied:
Got it.

mfeurer reviewed (commented):
I just skimmed this and left a few comments. I think this is a good idea. Do you plan to add a build action to github actions?

PGijsbers commented on docker/readme.md:
```diff
@@ -0,0 +1,78 @@
+It works similar to building docs, but should specify `test` as mode.
+For example, to run tests on your local repository:
+
+    docker run --mount type=bind,source=PATH_TO_REPOSITORY,destination=/code -t openml/openml-python test
```
Yes, added clarification.

PGijsbers commented:
I added a Github workflow that automatically builds and pushes a docker image on a push to `develop` or `docker` (the latter is only for development purposes, but I figured it might come in handy later too so I did not remove it).
The only unresolved issue is that the image is always pushed with the `openml/openml-python:latest` tag only. I tried to also push with the `GITHUB_SHA` tag, but that wouldn't take for some reason. I can dive into the syntax for that another time, but I don't have time for it now.
Documentation is also updated. New links point to `main` instead of `master`.

mfeurer commented on CONTRIBUTING.md:
```diff
@@ -178,6 +178,10 @@ following rules before you submit a pull request:
  - If any source file is being added to the repository, please add the BSD 3-Clause license to it.
 
 
+note: *We recommend to follow the instructions below to install all requirements locally.
```
```suggestion
*Note*: We recommend to follow the instructions below to install all requirements locally.
```

mfeurer commented on CONTRIBUTING.md:
```diff
@@ -178,6 +178,10 @@ following rules before you submit a pull request:
 
+note: *We recommend to follow the instructions below to install all requirements locally.
+However it is also possible to use the [openml-python docker image](https://github.com/openml/openml-python/blob/main/docker/readme.md) for testing and building documentation.
+This can be useful for one-off contributions or when you are experiencing installation issues.*
```
```suggestion
This can be useful for one-off contributions or when you are experiencing installation issues.
```

mfeurer reviewed (commented):
This looks good, interestingly, the workflow is a bit different to the one we use in Auto-sklearn (https://github.com/automl/auto-sklearn/blob/master/.github/workflows/docker-publish.yml). I think we can copy over one or two things, such as running the unit tests and correctly setting the tag, but besides that it looks good to me.
